src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProviderFields.cs

```diff
@@ -38,50 +39,101 @@ public ModelTypeProviderFields(InputModelType inputModel, TypeFactory typeFactor
+                var parameter = new Parameter(parameterName, BuilderHelpers.EscapeXmlDocDescription(inputModelProperty.Description), field.Type, null, parameterValidation, null);
                 parametersToFields[parameter.Name] = field;
                 // all properties should be included in the serialization ctor
-                serializationParameters.Add(parameter);
-                // only required + not readonly + not literal property + not discriminator could get into the public ctor
-                if (inputModelProperty.IsRequired &&
-                    !inputModelProperty.IsReadOnly &&
-                    !inputModelProperty.IsDiscriminator &&
-                    inputModelProperty.Type is not InputLiteralType)
+                serializationParameters.Add(parameter with { Validation = Validation.None });
+
+                // for classes, only required + not readonly + not property with constant value + not discriminator could get into the public ctor
+                // for structs, all properties must be set in public constructor
+                if (isStruct || inputModelProperty is { IsRequired: true, IsDiscriminator: false, ConstantValue: null })
                 {
-                    publicParameters.Add(parameter);
+                    if (Configuration.Generation1ConvenienceClient || !inputModelProperty.IsReadOnly)
+                    {
+                        publicParameters.Add(parameter with { Type = TypeFactory.GetInputType(parameter.Type) });
+                    }
                 }
             }
 
+            if (additionalPropertiesType is not null)
+            {
+                // We use a $ prefix here as AdditionalProperties comes from a swagger concept
+                // and not a swagger model/operation name to disambiguate from a possible property with
+                // the same name.
+                var existingMember = sourceTypeMapping?.GetMemberByOriginalName("$AdditionalProperties");
+
+                var type = typeFactory.CreateType(additionalPropertiesType);
+                if (!inputModelUsage.HasFlag(InputModelTypeUsage.Input))
+                {
+                    type = TypeFactory.GetOutputType(type);
+                }
+
+                var name = existingMember is null ? "AdditionalProperties" : existingMember.Name;
+                var declaration = new CodeWriterDeclaration(name);
+                declaration.SetActualName(name);
+
+                var accessModifiers = existingMember is null ? Public : GetAccessModifiers(existingMember);
+
+                var additionalPropertiesField = new FieldDeclaration($"Additional Properties", accessModifiers | ReadOnly, type, declaration, null, false, true);
+                var additionalPropertiesParameter = new Parameter(name.ToVariableName(), "Additional Properties", type, null, Validation.None, null);
```

Comment:
I am always curious why we have to define a field declaration first before we build a property.
We have to call the `SetActualName` to resolve the issue of when the CodeWriterDeclaration gets written.
Does it really worth it? Could we just simply build `ObjectTypeProperty` here and return them?

Reply:
From what I recall, there was a real case when there was a conflict of field names. We can return back to this after MPG models are switched to the `InputModelType`